### ISSUE TITLE
A005 stdlib-module-shadowing warns that utils.logging shadows python logging module

### ISSUE THREAD

**Issue by dwt** [astral-sh/ruff#15399]
[This rule](https://docs.astral.sh/ruff/rules/stdlib-module-shadowing/) warns for me that `utils.logging` shadows stdlib `logging`.

Is this intentional? The whole point why our logging module is in `utils` is so the names do not collide. Am I missing something?

**Comment by MichaReiser** [comment]
Hi. Thanks for opening the question. 

This seems to be matching the upstream behavior (more as a note to myself rather than this being a justification why):

https://github.com/gforcada/flake8-builtins/blob/5484162ace8d9546513be710d70e9c148010d9cc/flake8_builtins.py#L292

I'm not 100% sure if this is the right behavior or if the rule is too eager to warn. IMO, the rule should warn if your module is named `logging` because it then shadows the stdlib module. Naming it `utils.logging` seems less problematic to me. @AlexWaygood am I overlooking something?

**Comment by mattmess1221** [comment]
I think this rule should only trigger for modules if there isn't an `__init__.py` file in the directory OR any module in the package is executable. e.g. it has a shebang or `if __name__ == '__main__':` check.

This is the only way I can see module shadowing being possible, as doing so prepends the parent directory to `sys.path`, overriding the stdlib.

**Comment by MichaReiser** [comment]
I discussed this with @AlexWaygood in our 1:1 and we think that we should change the default to only flag modules that have the exact same full qualified name as a standard library module and instead, add an option to enable the more strict linting that flags modules that only have the same name. 

The motivation for an option is that some projects might want stricter behavior to guarantee that modules don't shadow builtins when running a module from a subdirectory (e.g. `cd` into the `utils` directory and run a script from there).

**Comment by martina-oefelein** [comment]
It triggers if `util` is a regular package (contains an `__init__.py`):
```
.
├── pyproject.toml
└── src
    └── test
        ├── __init__.py
        └── utils
            ├── __init__.py
            └── logging.py
```

```bash
$ uvx ruff check --isolated --select A005 src
src/test/utils/logging.py:1:1: A005 Module `logging` shadows a Python standard-library module
```

**Comment by scott-boost** [comment]
```.
├── pyproject.toml
└── AcmeOrg
    └── http
        ├── __init__.py
        └── utils
            ├── __init__.py
            └── logging.py
```
interesting that flake8 didnt trigger A005 in this setup BUT ruff did
(notice that AcmeOrg doesnt contain a `__init__.py` file)

**Comment by rh-sp** [comment]
### Opinion

I think that the current rule is too strict, but "only fully qualified names" sounds like it could be too loose, depending on how it's interpreted. In my opinion, any root-level packages that are named the same as any stdlib root-level package should be blocked, _along with all their child packages._

### Example

- ~~abc~~ ❌
- ~~collections~~ ❌
  - ~~abc~~ ❌
  - ~~foobar~~ ❌
- foobar ✅
  - abc ✅
  - collections ✅
    - abc ✅
- urlparse ✅

### Reasoning

I haven't been around then, but from what I've heard, in the olden days, it wasn't well-defined what `import abc` does when in a package. Does it refer to the root-level package `abc`, or the `abc` defined at the current package level? So in that sense, for older Pythons, it made sense to flag `foobar.abc` as potentially shadowing `abc`.

However, in modern Python, this issue has been resolved. `import abc` _always_ refers to the root-level package, _never_ `foobar.abc` (or `collections.abc`). At the same time, `.abc` (which can stand for `foobar.abc` or `collections.abc`, depending on context) can _never_ shadow `abc`. So, `foobar.abc` does not need to trigger the rule.

Additionally, `collections.foobar` should _also_ trigger the rule, even though the fully qualified name doesn't exist in `stdlib`. This is because if you `import collections.foobar`, while the import statement itself can only mean one thing, from what I understand in order for it to work, `collections` _must_ be shadowed, so the problem is still there.

Hope that makes sense :)

**Comment by flying-sheep** [comment]
> `import abc` _always_ refers to the root-level package, _never_ `foobar.abc` (or `collections.abc`). At the same time, `.abc` (which can stand for `foobar.abc` or `collections.abc`, depending on context) can _never_ shadow `abc`. So, `foobar.abc` does not need to trigger the rule.

Yeah definitely. This rule has a lot of false positives.

**Comment by dwt** [comment]
🎉

**Comment by glopesdev** [comment]
@ntBre We are having the opposite problem where our CI started emitting more warnings following the latest release of ruff, in places where it wasn't emitting before.

Is there documentation somewhere explaining how exactly to turn on the new "default" behavior?

**Comment by glopesdev** [comment]
For those randomly checking here, the new config for `pyproject.toml` should be:

```
[tool.ruff]
lint.flake8-builtins.builtins-strict-checking = false
```

Still not clear to me why an issue which was supposed to reduce the number of default built-in warnings ended up producing more warnings under the default configuration.

**Comment by ntBre** [comment]
It will be a breaking change to update the default, so that will have to wait for the next minor version (0.10.0). In preview mode, the default should already be `lint.flake8-builtins.builtins-strict-checking = false`.

But yes, as part of this change there was also a bug fix to bring our implementation of A005 in line with the upstream, which *does* produce more warnings in some cases now.

**Comment by ntBre** [comment]
Oh and yes, that was my mistake on the documentation. I have an open PR (#16097) to make that easier to find that will hopefully be merged today!
